### PR TITLE
feat: added option to modify icebreakers with ai

### PR DESCRIPTION
**Pull request by BartoszJarocki**
# Description

Implements a terrible idea about modifying icebreakers with AI. [Thread 🔒 ](https://parabol.slack.com/archives/C03NGEJUKCN/p1701807407413389)

## Demo

CleanShot 2023-12-06 at 15 29 56@2x

## Testing scenarios

- [ ] Make sure only facilitator can modify icebreakers
- [ ] Make sure modified icebreakers are only visible to others upon approval by facilitator

## Final checklist

- [x] I checked the [code review guidelines](../docs/codeReview.md)
- [x] I have added [Metrics Representative](../docs/codeReview.md#metrics-representative) as reviewer(s) if my PR invovles metrics/data/analytics related changes
- [x] I have performed a self-review of my code, the same way I'd do it for any other team member
- [x] I have tested all cases I listed in the testing scenarios and I haven't found any issues or regressions
- [x] Whenever I took a non-obvious choice I added a comment explaining why I did it this way
- [x] I added the label `One Review Required` if the PR introduces only minor changes, does not contain any architectural changes or does not introduce any new patterns and I think one review is sufficient'
- [x] PR title is human readable and could be used in changelog

### CODE REVIEW

File: packages/server/graphql/public/mutations/modifyCheckInQuestion.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  if (endedAt) {
+    return standardError(new Error('Meeting has already ended'), {userId: viewerId})
+  }
+
+  const openai = new OpenAIServerManager()
+  const modifiedCheckInQuestion = await openai.modifyCheckInQuestion(checkInQuestion, modifyType)
+
+  analytics.icebreakerModified(viewerId, meetingId, modifyType)
```
Comment (BartoszJarocki) [inline packages/server/graphql/public/mutations/modifyCheckInQuestion.ts:41]: @tianrunhe is that enough to track this feature?

**Review by tianrunhe — APPROVED** [review]
LGTM from metrics side!

**Review by tianrunhe — APPROVED** [review]
LGTM from metrics side!

File: packages/server/utils/analytics/analytics.ts [hunk 2]
```diff
@@ -163,6 +164,7 @@ export type AnalyticsEvent =
   | 'Conversion Modal Pay Later Clicked'
   // kudos
   | 'Kudos Sent'
+  | 'Icebreaker modified'
```
[the remark this answers is not in the document]
Comment (BartoszJarocki) [reply packages/server/utils/analytics/analytics.ts:167]: good catch!

**Review by tianrunhe — CHANGES_REQUESTED** [review]

File: packages/server/graphql/public/mutations/modifyCheckInQuestion.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (BartoszJarocki) [reply packages/server/graphql/public/mutations/modifyCheckInQuestion.ts:41]: ah yes, OpenAI API may break so then it'll return null. It's ok, frontend handles that but from metrics perspective, you're right!

[the remark this answers is not in the document]
Comment (BartoszJarocki) [reply packages/server/graphql/public/mutations/modifyCheckInQuestion.ts:41]: @tianrunhe after giving it another thought, are you sure we want to send this even only when open ai returns something? I think it's more important to know that someone actually tried to use the feature, not that open ai call was successful. no strong opinions though!

[the remark this answers is not in the document]
Comment (tianrunhe) [reply packages/server/graphql/public/mutations/modifyCheckInQuestion.ts:41]: Good point! How about an additional property, maybe `success`, in the metric to indicate if we generate a new question for the user? Like what we did [in `Task Estimate Set`](https://github.com/ParabolInc/parabol/blob/8520063c7835bb5a685e83eabd735ee694d76486/packages/server/utils/analytics/analytics.ts#L85)

[the remark this answers is not in the document]
Comment (BartoszJarocki) [reply packages/server/graphql/public/mutations/modifyCheckInQuestion.ts:41]: that sounds good! will do

**Review by tianrunhe — APPROVED** [review]

File: packages/server/utils/OpenAIServerManager.ts [hunk 3]
```diff
@@ -243,6 +244,44 @@ class OpenAIServerManager {
       return null
     }
   }
+
+  async modifyCheckInQuestion(question: string, modifyType: ModifyType) {
+    if (!this.openAIApi) return null
+
+    const prompt: Record<ModifyType, string> = {
+      EXCITING: `Transform the following team retrospective ice breaker question into something imaginative and unexpected, using simple and clear language suitable for an international audience. Keep it engaging and thrilling, while ensuring it's easy to understand.
```
Comment (mattkrick) [inline packages/server/utils/OpenAIServerManager.ts:252]: -1 either this needs a length limit, or we need to adjust the UI to go wider. if you expand the image below you'll see it starts to shove the listen icon under the meeting bar

Screenshot 2023-12-08 at 2 11 12 PM

**Review by mattkrick — CHANGES_REQUESTED** [review]
looking great! love the button component, questions can get a lil long, but great besides that

File: packages/client/ui/Button/Button.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import clsx from 'clsx'
+
+type Variant = 'primary' | 'secondary' | 'destructive' | 'ghost' | 'link' | 'outline'
+type Size = 'sm' | 'md' | 'lg' | 'default'
+type Shape = 'pill' | 'circle' | 'default'
+
+const BASE_STYLES =
+  'cursor-pointer inline-flex items-center justify-center whitespace-nowrap font-semibold transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50'
+
+// TODO: make sure the styles match the designs
```
[the remark this answers is not in the document]
Comment (BartoszJarocki) [reply packages/client/ui/Button/Button.tsx:12]: I only made sure buttons I use are good to go so I'd leave it. I planned to write on engineering channel about this and replacing all the styled components and copy/paste tailwind buttons with this while ensuring everything looks good. how does it sound? 

I'll do my best to at least try to make it good for every type of the button, though. how does it sound?

File: packages/server/graphql/public/typeDefs/modifyCheckInQuestion.graphql [hunk 5]
```diff
@@ -0,0 +1,25 @@
+enum ModifyType {
```
[the remark this answers is not in the document]
Comment (BartoszJarocki) [reply packages/server/graphql/public/typeDefs/modifyCheckInQuestion.graphql:1]: I'd leave it as is and not try to abstract too early. it won't be a hard change and we don't even know if we're gonna like the feature so far

File: packages/client/modules/meeting/components/MeetingCheckInPrompt/NewCheckInQuestion.tsx [hunk 6]
```diff
@@ -146,53 +167,177 @@ const NewCheckInQuestion = (props: Props) => {
...
+    const contentState = editorState.getCurrentContent()
+    const jumpToEnd = (selection as any).merge({
+      anchorOffset: contentState.getLastBlock().getLength(),
+      focusOffset: contentState.getLastBlock().getLength()
+    }) as SelectionState
+    const nextEditorState = EditorState.forceSelection(editorState, jumpToEnd)
+    setEditorState(nextEditorState)
+  }
+
+  const {
```
[the remark this answers is not in the document]
Comment (BartoszJarocki) [reply packages/client/modules/meeting/components/MeetingCheckInPrompt/NewCheckInQuestion.tsx]: omg how could I miss that. old habits die hard, fixed!

File: packages/server/utils/OpenAIServerManager.ts (hunk 3 above)
Comment (BartoszJarocki) [reply packages/server/utils/OpenAIServerManager.ts]: really good point. added char limit to the prompt and made the layout a little bit compact. modified questions are even better now, thank you!

**Comment by BartoszJarocki** [comment]
@mattkrick ready for another round!

File: packages/client/ui/Button/Button.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattkrick) [reply packages/client/ui/Button/Button.tsx:12]: sounds good!

**Review by mattkrick — APPROVED** [review]

**Pull request merged**